RTS Formations Example

Ji-Rath opened the pull request:
Apologies in advance for the duplicate PR, just goofed up with my fork and made a PR in main
Basically, just a simple PR to add a formations example to MassSample. In my opinion, its not breaking any barriers, but I think it's a good example to see what is possible with Mass. It is all encapsulated in its own plugin too! Consider checking out #37 for more info.

https://user-images.githubusercontent.com/1747157/179067289-21b201ce-74e0-4a17-ae5c-e3d0bafff4ee.mp4

Ji-Rath committed:
Add RTSFormationsPlugin

Ji-Rath committed:
Some small discoveries about LOD Processors

Ji-Rath committed:
Fix crash on endplay
Adjust entity position in formation to try and make it more seamless - WIP

Ji-Rath committed:
Second pass on unit formations. While it is more organized than previously, it can use more work. Entities now go to location and choose a destination based on distance.

Ji-Rath committed:
Begin optimizing new formation system

Ji-Rath committed:
Add simple variable tick rate to rts movement processor

Ji-Rath committed:
Performance refactoring

Optimize and cleanup rts subsystem for large-scale use
Still need to consider if its feasible to move some logic to a processor rather than the subsystem

Ji-Rath committed:
Fix single entity not moving when changing position occasionally
Get units to move in formation...at least good enough
Adjust tolerance for reaching formation position

Ji-Rath committed:
Fix entity creation/destruction not properly updating unit
Add RTSUpdateEntityIndex to spread out logic cost when finding formation index
Fix bug where entities would stop moving when travelling to new destination

Ji-Rath committed:
Implement entity offset so that unit position doesnt have to be recalculated every processor tick. This also allows me to implement different formations in one spot.

Ji-Rath committed:
Refactor position logic for units to support hollow formation

Ji-Rath committed:
Implement basic circle formation

Ji-Rath committed:
Add formation presets

Ji-Rath committed:
Add circle formation variables
Move enum formation to FormationPresets to avoid circular dependency, should be moved to its own class in the future

Ji-Rath committed:
Attempt to resolve incorrect order while smooth rotating unit
Implement smooth rotation of unit when the angle is not extreme

Ji-Rath committed:
Refactor formation rotation logic to use FRotator since the logic for interpolation is already built-in
Temporarily remove optimizations for debugging
Add specific logic for smooth rotations at close angles while recalculating units index when making large turns
Remove LerpDegrees as it is no longer used

Ji-Rath committed:
Refactor UpdateUnitPosition and implement CalculateNewPositions() to better separate code chunks
Implement simple comments to core functions in RTSFormationSubsystem for my own sanity

Ji-Rath committed:
Cleanup RTSFormationSubsystem
Modify turn rotation on units so they turn about half as fast - should be a variable in the future

Ji-Rath committed:
Extend the Gameplay Debuggger to enable RTS Unit debugging
Add GameplayDebuggerCategory_RTSAgents
Add GameplayDebugger to dependency list

Ji-Rath committed:
Add RTSAgentSubsystem for agent position tracking and communication to mass processors

Ji-Rath committed:
Add LaunchEntityProcessor to handle launching entities

Ji-Rath committed:
Begin implementing HashGrid to entities for querying positions

Ji-Rath committed:
Implement basic hashgrid processors to update and set location of entities

Megafunk commented:
Sorry about the lack of response, I'll get around to looking at this over the weekend!

Ji-Rath commented:
No problem! I got a little carried away with some of the functionality (and some WIP features) so let me know if it is too complex or should be condensed to fit the MassSample.

Megafunk commented:
This is sweet because it's an often requested feature (RTS stuff!) It's even in a separate plugin which shows off the power of composition and saves people who don't need it some space. 

However, I think I broke it? 

https://user-images.githubusercontent.com/6806463/188527664-3521f468-819d-444d-914b-54c1fe4af181.mp4

Did I miss something? I assume there is some setup I forgot or I may be using the wrong pawn. Do you have some content only stored in your project or something?

Ji-Rath commented:
> This is sweet because it's an often requested feature (RTS stuff!) It's even in a separate plugin which shows off the power of composition and saves people who don't need it some space.
> 
> However, I think I broke it?
> 
>  UnrealEditor-Win64-DebugGame_2022-09-05_19-23-36.mp4 
> Did I miss something? I assume there is some setup I forgot or I may be using the wrong pawn. Do you have some content only stored in your project or something?

Ah shucks, Im going to take another look and get back to you on that. I think there might be some content I forgot to move over to the plugin.

Ji-Rath committed:
Clean up Signaling entities and ensure that there are entities to signal before calling
Fix units not updating their position correctly when setting a new preset or being spawned in-game

Ji-Rath committed:
Fix entities not updating their index due to legacy code
Move Hash grid processors to their own file

Ji-Rath committed:
Add RTSAgentProcessors to handle hash grid positions of RTS agents

Ji-Rath committed:
Change LaunchEntityProcessor to a Signal Processor since I couldnt get an observer working for the life of me - should be further investigated

Ji-Rath committed:
Add BP_RTSPlayer for birds-eye view of map
Add BP_Pad for base class of interactables

Ji-Rath committed:
Modify BP_SpawnUnit and BP_SpawnEntity to use new base class BP_Pad

Ji-Rath committed:
Cleanup BP_FormationGameMode and re-enable randomized position generating

Ji-Rath committed:
Modify DA_AgentFormation so LODSimulation is no longer included by default, will need to determine whether its worth it to create an LODCollectorProcessor or leave it out for simplicity

Ji-Rath committed:
Adjust LaunchEntities function to use QuerySmall and Signal processing due to roadblock with observers - needs further investigation

Ji-Rath committed:
Update RTSFormationExample map so that updated gamemode and character are included in the plugin

Ji-Rath commented:
The current version still has a few bugs to iron out but all my changes should be available. Let me know if for some reason the map is still not functional. The gamemode in the map should be `BP_FormationGameMode` if its not already set 👍 
https://user-images.githubusercontent.com/1747157/189175480-dfe58d99-092d-4122-b105-ec6d4d65ca61.mp4

Ji-Rath committed:
Fix header files being in wrong folder location

Megafunk commented:
The fixes worked! I can now make units that run around in formation!

As for the max of 100 issue mentioned in the original pr my assumption is it's just the LODMax count but I'm not seeing anything wacky happening so far.
![image](https://user-images.githubusercontent.com/6806463/190889108-5d2e3efe-7b2c-4297-b61c-890759656cc9.png)

I'm seeing some exceptions triggered from actor representation when clicking on actors at times but this is definitely good enough to include as "experimental" in my mind. Nice!!

Ji-Rath committed:
Merge branch 'Megafunk:main' into RTSFormations

Ji-Rath committed:
Update to 5.4
Resolve some rare crashes

Ji-Rath committed:
Update how subsystems are retrieved so they are more inline with native methods

Ji-Rath committed:
Push RTSGameMode to support simple top down view

LangArthur commented:
Hi! I just found this pr, while I want to work on a rts type game. I looked at the code and it seems to be a great example of what I want to achieve.

However, does the video you showed in this pr came from the NewMap level? Because I cannot manage to open it with 5.4, and I tried to create a new level on my own but this is not yet a success (I am quite new to unreal so I spend ages to do simple things :D).

For the completeness here is the only warning I have when loading the level:
```
SmartObjectCollection_0 /Game/NewMap.NewMap:PersistentLevel.SmartObjectCollection_0 : Obsolete and must be removed! (Class is deprecated) 
```

Ji-Rath commented:
> Hi! I just found this pr, while I want to work on a rts type game. I looked at the code and it seems to be a great example of what I want to achieve.
> 
> However, does the video you showed in this pr came from the NewMap level? Because I cannot manage to open it with 5.4, and I tried to create a new level on my own but this is not yet a success (I am quite new to unreal so I spend ages to do simple things :D).
> 
> For the completeness here is the only warning I have when loading the level:
> 
> ```
> SmartObjectCollection_0 /Game/NewMap.NewMap:PersistentLevel.SmartObjectCollection_0 : Obsolete and must be removed! (Class is deprecated) 
> ```

I have a similar version hosted on my repo [MassAIExample](https://github.com/Ji-Rath/MassAIExample). It should be updated to the latest version of UE. Let me know if you have any questions. Do keep in mind this is simply an example and I highly recommend reading the code to get a better idea of the logic.

LangArthur commented:
Thanks for the answer ! I'll check your repo. And of course I checked the code first in this pr, but I'm a bit confused about the seperation between what you have done in the editor side and in the code side and how you retrieve all the entities created by the spawner.

In my project, I want to do everything programmatically, since the loaded level will be dynamic.

If I do have more concrete question, I'll surely open an issue on your MassAiExample repo, to avoid spamming here ;)
